galsim/wfirst/__init__.py

```diff
@@ -0,0 +1,188 @@
+#    file.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions, and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+#
+"""
+The galsim.wfirst module, containing information GalSim needs to simulate images for the WFIRST-AFTA
+project.
+
+This module contains numbers and routines for the WFIRST-AFTA project.  There is also a demo
+illustrating the use of most of this functionality, demo13.py.  Some of the parameters below relate
+to the entire wide-field imager.  Others, especially the return values of the functions to get the
+PSF and WCS, are specific to each SCA (Sensor Chip Assembly, the equivalent of a chip for an optical
+CCD) and therefore are indexed based on the SCA.  All SCA-related arrays are 1-indexed, i.e., the
+entry with index 0 is None and the entries from 1 to n_sca are the relevant ones.  This is
+consistent with diagrams and so on provided by the WFIRST project, which are 1-indexed.
+
+Currently, the module includes the following numbers:
+
+    gain - The gain for all SCAs (sensor chip assemblies) is expected to be the same, so this is a
+           single value rather than a list of values.
+
+    pixel_scale - The pixel scale in units of arcsec/pixel.
+
+    diameter - The telescope diameter in meters.
+
+    obscuration - The linear obscuration of the telescope, in terms of fraction of the diameter.
+
+    exptime - The typical exposure time in units of seconds.  The number that is stored is for a
+              single dither.  Each location within the survey will be observed with a total of 5-7
+              dithers across 2 epochs.  For a single dither, there are 32 up-the-ramp samples, each
+              taking 5.423 seconds, but the effective live time is really 31 samples.  This is the
+              source of the 168.1s that is currently stored for exptime.
+
+    n_dithers - The number of dithers per filter (typically 5-7, so this is currently 6 as a
+                reasonable effective average).
+
+    dark_current - The dark current in units of e-/pix/s.
+
+    nonlinearity_beta - The coefficient of the (counts)^2 term in the detector nonlinearity
+                        function.  This will not ordinarily be accessed directly by users; instead,
+                        it will be accessed by the convenience function in this module that defines
+                        the nonlinearity function.
+
+    reciprocity_alpha - The normalization factor that determines the effect of reciprocity failure
+                        of the detectors for a given exposure time.  Typically, users would then
+                        simulate the reciprocity failure for a given image `im` and for the default
+                        WFIRST exposure time using
+
+        >>>> im_recip = im.addReciprocityFailure(galsim.wfirst.exptime,
+                                                 galsim.wfirst.reciprocity_alpha)
```

Comment:
Okay - done, including unit tests and putting them into demo13.  Please check out the last few commits.  @arunkannawadi , can you also see what you think of the new detector routines in the WFIRST module?

Two comments:
- I left the parameters that go into the detector effects in ```__init.py__```, instead of keeping them more hidden in ```wfirst_detectors.py```.  My reasoning was that if people want to query those values or read about them or try modified versions of them, it's kind of simpler if everything is together in one place.

- I did not address the suggestion to make a routine that does ALL detector effects in a proper sequence.  I really like the idea, but since I noticed there were some comments on how the detector effects were done in demo13, I want to address those before making that routine.  This suggestion will therefore stay on my list for a bit later.  (Also, my plan is to keep demo13 with all effects included explicitly, so people can read about the order there and see how it is done.  I would just put a comment that people who don't want to put all that code can use the ```galsim.wfirst.allDetectorEffects()``` routine, or whatever we end up calling it.)